### PR TITLE
Expose parameters_normalized param in invoke_workflow

### LINKED ISSUE #321: expose already_normalized to workflow invocation
When filling parameters in subworkflows, following error message occurs from galaxy.workflows.run_request._normalize_step_parameters():

`bioblend.ConnectionError: Unexpected HTTP status code: 400: {"err_msg": "Specifying subworkflow step parameters requires already_normalized to be specified as true.", "err_code": 400008}`

Perhaps add an "already_normalized" parameter to invoke_workflow()?  For my needs, simply changing False to True here works:

`already_normalized = payload.get('parameters_normalized', True)`

### PR DESCRIPTION
**Pull request by rikeshi**
Fix #321

I am not sure how to test this functionality because I do not know what kind of input would require this parameter to be set to True.

### CODE REVIEW

**Comment by simonbray** [comment]
@rikeshi the issue affects specifying params for subworkflows - I uploaded a workflow you can use for testing.

Try `gi.workflows.invoke_workflow(wf_id, inputs={'0': {'src': 'hda', 'id': dataset_id,}}, params={'1': {'1|2': 'comma'}})` and check it fails, then try with the new parameter.

**Comment by rikeshi** [comment]
Thanks, @simonbray

File: bioblend/_tests/TestGalaxyWorkflows.py [hunk 1]
```diff
@@ -64,6 +65,26 @@ def invocation_steps_by_order_index():
...
+                inputs={'0': {'src': 'hda', 'id': dataset_id}},
+                params={'1': {'1|2': 'comma'}}
+            )
+        invocation = self.gi.workflows.invoke_workflow(
+            workflow['id'],
+            inputs={'0': {'src': 'hda', 'id': dataset_id}},
+            params={'1': {'1|2': 'comma'}},
+            parameters_normalized=True
+        )
+        self.assertNotIn(invocation['state'], INVOCATION_TERMINAL_STATES)
```
[the remark this answers is not in the document]
Comment (rikeshi) [reply bioblend/_tests/TestGalaxyWorkflows.py]: No, you are correct that it serves no real purpose here.

**Review by simonbray — APPROVED** [review]

File: bioblend/galaxy/workflows/__init__.py [hunk 2]
```diff
@@ -555,7 +567,7 @@ def invoke_workflow(self, workflow_id, inputs=None, params=None,
           (which is stable across workflow imports) or the step UUID which is
           also stable.
         """
-        payload = {'workflow_id': workflow_id}
+        payload: Dict[str, Union[str, bool, dict]] = {'workflow_id': workflow_id}
```
Comment (nsoranzo) [inline bioblend/galaxy/workflows/__init__.py:570]:
```suggestion
        payload = {'workflow_id': workflow_id}
```

File: bioblend/galaxy/workflows/__init__.py [hunk 3]
```diff
@@ -447,6 +453,12 @@ def invoke_workflow(self, workflow_id, inputs=None, params=None,
         :param inputs_by: Determines how inputs are referenced. Can be
           "step_index|step_uuid" (default), "step_index", "step_id", "step_uuid", or "name".
 
+        :type parameters_normalized: bool
+        :param parameters_normalized: Whether Galaxy should normalize ``params``
+          to ensure everything is referenced by a numeric step ID. Default is
+          ``True``, but when setting ``params`` for a subworkflow, ``False`` is
```
Comment (nsoranzo) [inline bioblend/galaxy/workflows/__init__.py:459]:
```suggestion
          ``False``, but when setting ``params`` for a subworkflow, ``True`` is
```

File: bioblend/_tests/TestGalaxyWorkflows.py [hunk 4]
```diff
@@ -64,6 +64,25 @@ def invocation_steps_by_order_index():
         invocation = self.gi.workflows.show_invocation(workflow_id, invocation_id)
         self.assertEqual(invocation["state"], "scheduled")
 
+    @test_util.skip_unless_galaxy('release_19.01')
+    def test_invoke_workflow_parameters_normalized(self):
+        path = test_util.get_abspath(os.path.join('data', 'paste_columns_subworkflow.ga'))
+        workflow = self.gi.workflows.import_workflow_from_local_path(path)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyWorkflows.py:70]:
```suggestion
        workflow_id = self.gi.workflows.import_workflow_from_local_path(path)["id"]
```

File: bioblend/_tests/TestGalaxyWorkflows.py [hunk 5]
```diff
@@ -64,6 +64,25 @@ def invocation_steps_by_order_index():
...
+        history_id = self.gi.histories.create_history(name="TestWorkflowInvokeParametersNormalized")["id"]
+        dataset_id = self._test_dataset(history_id)
+        with self.assertRaises(ConnectionError):
+            self.gi.workflows.invoke_workflow(
+                workflow['id'],
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyWorkflows.py:75]:
```suggestion
                workflow_id,
```

File: bioblend/_tests/TestGalaxyWorkflows.py [hunk 6]
```diff
@@ -64,6 +64,25 @@ def invocation_steps_by_order_index():
...
+                inputs={'0': {'src': 'hda', 'id': dataset_id}},
+                params={'1': {'1|2': 'comma'}}
+            )
+        self.gi.workflows.invoke_workflow(
+            workflow['id'],
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyWorkflows.py:80]:
```suggestion
            workflow_id,
```

File: bioblend/galaxy/workflows/__init__.py (hunk 2 above)
Comment (rikeshi) [reply bioblend/galaxy/workflows/__init__.py:570]: Without a type hint here linting fails with the following errors:

```
GLOB sdist-make: /home/fread/uni/ws2021/master-project/bioblend/setup.py
lint inst-nodeps: /home/fread/uni/ws2021/master-project/bioblend/.tox/.tmp/package/1/bioblend-0.15.0.zip
lint installed: attrs==20.3.0,bioblend @ file:///home/fread/uni/ws2021/master-project/bioblend/.tox/.tmp/package/1/bioblend-0.15.0.zip,boto==2.49.0,certifi==2020.12.5,chardet==4.0.0,flake8==3.9.0,flake8-bugbear==21.4.3,flake8-import-order==0.18.1,idna==2.10,mccabe==0.6.1,mypy==0.812,mypy-extensions==0.4.3,pycodestyle==2.7.0,pyflakes==2.3.1,PyYAML==5.4.1,requests==2.25.1,requests-toolbelt==0.9.1,typed-ast==1.4.2,typing-extensions==3.7.4.3,urllib3==1.26.4
lint run-test-pre: PYTHONHASHSEED='1274784429'
lint run-test: commands[0] | flake8 .
lint run-test: commands[1] | mypy bioblend/
bioblend/galaxy/workflows/__init__.py:570:33: error: Incompatible types in assignment (expression has type "Dict[Any, Any]", target has type "str")
bioblend/galaxy/workflows/__init__.py:573:37: error: Incompatible types in assignment (expression has type "Dict[Any, Any]", target has type "str")
bioblend/galaxy/workflows/__init__.py:576:45: error: Incompatible types in assignment (expression has type "Dict[Any, Any]", target has type "str")
bioblend/galaxy/workflows/__init__.py:583:44: error: Incompatible types in assignment (expression has type "bool", target has type "str")
bioblend/galaxy/workflows/__init__.py:585:55: error: Incompatible types in assignment (expression has type "bool", target has type "str")
bioblend/galaxy/workflows/__init__.py:589:48: error: Incompatible types in assignment (expression has type "bool", target has type "str")
Found 6 errors in 1 file (checked 65 source files)
```

Comment (rikeshi) [reply bioblend/galaxy/workflows/__init__.py:570]: Just `payload: dict =`  also works.

Comment (nsoranzo) [reply bioblend/galaxy/workflows/__init__.py]: Thanks, this stuff is also new for me! I've pushed a commit to use the `Dict[str, Any]` annotation, which is kind of a middle ground.

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
Thanks @rikeshi !

**Comment by rikeshi** [comment]
Thanks for reviewing @simonbray and @nsoranzo !

**Pull request merged**